File: aoc24/d2/asolu.py
X = [list(map(int, x.split(" "))) for x in open("input.txt", "r").read().splitlines()]

def diff(x):
    return [x1-x2 for x1,x2 in zip(x[:-1], x[1:])]

def test(seq):
    if all([e != 0 for e in seq]) and all([0<abs(e)<4 for e in seq]):
        return all(seq[c-1]/abs(seq[c-1]) == seq[c]/abs(seq[c]) for c in range(1,len(seq)))
    return False

res = 0
for x in X:
    if test(diff(x)):
        res += 1
        continue
    for j in range(len(x)):
        seq = diff(x[:j]+x[j+1:])
        if test(seq):
            res += 1
            break

print(res)

File: aoc24/d2/asolu.hs
import System.IO
import Control.Monad (forM_)
import Data.List (tails)

-- Function to calculate differences between consecutive elements
diff_ :: [Int] -> [Int]
diff_ xs = zipWith (-) xs (tail xs)

-- Function to test the sequence
test_ :: [Int] -> Bool
test_ seq = all (/= 0) seq && all (\e -> abs e < 4) seq && all (\c -> signum (seq !! (c - 1)) == signum (seq !! c)) [1..length seq - 1]

-- Function to check a sequence against the conditions
checkSequence :: [Int] -> Bool
checkSequence xs = test_ (diff_ xs) || any (test_ . diff_) (init $ tails xs)

main :: IO ()
main = do
    -- Read input from file
    contents <- readFile "input.txt"
    let xss = map (map read . words) (lines contents) :: [[Int]]

    -- Calculate the result
    let res = length $ filter id $ map checkSequence xss

    -- Print the result
    print res

